File: src/main/java/com/dalolorn/sr2modmanager/view/ModInstallerPane.java
package com.dalolorn.sr2modmanager.view;

import com.dalolorn.sr2modmanager.adapter.ModInstaller;
import com.dalolorn.sr2modmanager.adapter.Recommendation;
import javafx.application.Platform;
import javafx.collections.ObservableList;
import javafx.concurrent.Task;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.*;
import javafx.scene.layout.GridPane;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.List;

public class ModInstallerPane extends GridPane {
	@FXML Button connectButton;
	@FXML Button installButton;
	@FXML TextField urlField;
	@FXML Label urlLabel;
	@FXML Label recommendationLabel;
	@FXML TextArea modInfo;
	@FXML TextArea branchInfo;
	@FXML ListView<String> branchList;
	@FXML ListView<String> recommendationList;

	@FXML
	public void initialize() {
		branchList.getSelectionModel().selectedItemProperty().addListener((observable, oldValue, newValue) -> setActiveBranch(newValue));
		recommendationList.getSelectionModel().selectedIndexProperty().addListener((observable, oldValue, newValue) -> urlField.setText(newValue.intValue() >= 0 ? recommendationList.getItems().get(newValue.intValue()) : ""));

		try {
			Recommendation.load();
			recommendationList.getItems().addAll(Recommendation.getInstance().getRecommendationList());
		} catch (IOException e) {
			var msg = new Alert(Alert.AlertType.WARNING, "Could not load history.json! Cause:"  + e);
			e.printStackTrace();
			msg.showAndWait();
		}
	}

	private void setActiveBranch(@NotNull String branchName) {
		branchInfo.setText(ModInstaller.setActiveBranch(branchName));
		installButton.setDisable(!ModInstaller.hasBranch());
	}

	@FXML private void connectToRepository(ActionEvent event) {
		MainController.executeTask("Loading repository...", taskUpdateStage -> new Task<>() {
			@Override
			protected Void call() {
				String repoURL;
				ModInstaller.TextHandler errorHandler = error -> Platform.runLater(() -> {
					var msg = new ResizableAlert(Alert.AlertType.ERROR, error);
					msg.show();
				});

				if (
						(repoURL = ModInstaller.connectToRepository
								(
										urlField.getText(),
										this::updateMessage,
										errorHandler
								)
						) == null
				) return null;

				updateMessage("Fetching branch list...");
				if(!getBranches(errorHandler)) return null;

				updateMessage("Getting mod description...");
				if(!getDescription(errorHandler)) return null;

				Platform.runLater(() -> {
					setUrlText("Connected to " + repoURL);
					try {
						Recommendation.getInstance().addItem(urlField.getText());
						recommendationList.getItems().setAll(Recommendation.getInstance().getRecommendationList());
					} catch (IOException e) {
						var msg = new Alert(Alert.AlertType.WARNING, "Could not save history.json! Cause:"  + e);
						e.printStackTrace();
						msg.showAndWait();
					}
				});
				return null;
			}

			@Override
			protected void succeeded() {
				taskUpdateStage.close();
			}
		});
	}

	@FXML private void installMod(ActionEvent event) {
		MainController.executeTask("Preparing to install mod...", taskInstallStage -> new Task<>() {
			@Override
			protected Void call() {
				updateMessage("Preparing to install mod...");
				ModInstaller.installMod(
						warning -> Platform.runLater(() -> {
							Alert msg = new ResizableAlert(Alert.AlertType.WARNING, warning);
							msg.show();
						}),
						this::updateMessage,
						info -> Platform.runLater(() -> {
							Alert msg = new ResizableAlert(Alert.AlertType.INFORMATION, info);
							msg.show();
						}),
						error -> Platform.runLater(() -> {
							Alert msg = new ResizableAlert(Alert.AlertType.ERROR, error);
							msg.show();
						})
				);
				return null;
			}

			@Override
			protected void succeeded() {
				taskInstallStage.close();
			}
		});
	}

	boolean getBranches(ModInstaller.TextHandler errorHandler) {
		List<String> tagNames = ModInstaller.getBranches(errorHandler);
		if(tagNames == null) return false;

		Platform.runLater(() -> {
			branchList.getItems().clear();
			branchList.getItems().addAll(tagNames);
		});
		return true;
	}

	boolean getDescription(ModInstaller.TextHandler errorHandler) {
		String description = ModInstaller.getDescription(errorHandler);
		if(description == null) return false;

		Platform.runLater(() -> {
			modInfo.setText(description);

			branchInfo.setText("No branches selected");
		});
		return true;
	}

	void setUrlText(String url) {
		urlLabel.setText(url);
	}

	void deleteRepository() {
		var dialog = new ResizableAlert(Alert.AlertType.CONFIRMATION, "Are you sure you want to delete this repository from your computer? You will not be able to access it again until you redownload it.");
		ObservableList<ButtonType> dlgButtons = dialog.getDialogPane().getButtonTypes();
		dlgButtons.clear();
		dlgButtons.add(ButtonType.YES);
		dlgButtons.add(ButtonType.NO);
		dialog.setHeaderText("Delete Repository?");

		dialog.showAndWait()
				.filter(response -> response == ButtonType.YES)
				.ifPresent(response -> {
					if(ModInstaller.deleteRepository(error -> {
						var msg = new ResizableAlert(Alert.AlertType.ERROR, error);
						msg.show();
					})) {
						branchList.getItems().clear();
						modInfo.setText("");
						branchInfo.setText("");
					}
				});
	}

	void uninstallMod() {
		var dialog = new ResizableAlert(Alert.AlertType.CONFIRMATION, "Are you sure you want to uninstall this mod? The associated repository will not be removed from your computer.");
		ObservableList<ButtonType> dlgButtons = dialog.getDialogPane().getButtonTypes();
		dlgButtons.clear();
		dlgButtons.add(ButtonType.YES);
		dlgButtons.add(ButtonType.NO);
		dialog.setHeaderText("Uninstall Mod?");

		dialog.showAndWait()
				.filter(response -> response == ButtonType.YES)
				.ifPresent(response -> ModInstaller.uninstallMod(error -> {
					var msg = new ResizableAlert(Alert.AlertType.ERROR, error);
					msg.show();
				}));
	}
}

File: src/main/java/com/dalolorn/sr2modmanager/view/ModInstallerPane.kt
package com.dalolorn.sr2modmanager.view

import com.dalolorn.sr2modmanager.adapter.ModInstaller
import com.dalolorn.sr2modmanager.adapter.ModInstaller.TextHandler
import com.dalolorn.sr2modmanager.adapter.Recommendations
import javafx.application.Platform
import javafx.concurrent.Task
import javafx.fxml.FXML
import javafx.scene.control.*
import javafx.scene.control.Alert.AlertType
import javafx.scene.layout.GridPane
import javafx.stage.Stage
import java.io.IOException

class ModInstallerPane : GridPane() {
    @FXML var connectButton: Button? = null
    @FXML var installButton: Button? = null

    @FXML var urlField: TextField? = null
    @FXML var urlLabel: Label? = null
    @FXML var recommendationLabel: Label? = null

    @FXML var modInfo: TextArea? = null
    @FXML var branchInfo: TextArea? = null

    @FXML var branchList: ListView<String>? = null
    @FXML var modList: ListView<String>? = null

    @FXML var recommendationList: ListView<String>? = null

    @FXML
    fun initialize() {
        branchList!!.selectionModel.selectedItemProperty()
            .addListener { _, _, newValue ->
                setActiveBranch(newValue)
            }
        recommendationList!!.selectionModel.selectedIndexProperty()
            .addListener { _, _, newValue ->
                urlField!!.text = if (newValue.toInt() >= 0) recommendationList!!.items[newValue.toInt()] else ""
            }
        try {
            Recommendations.load()
            recommendationList!!.items.addAll(Recommendations.instance.recommendationList)
        } catch (e: IOException) {
            val msg = Alert(AlertType.WARNING, "Could not load history.json! Cause:$e")
            e.printStackTrace()
            msg.showAndWait()
        }
    }

    private fun setActiveBranch(branchName: String) {
        branchInfo!!.text = ModInstaller.setActiveBranch(branchName)
        installButton!!.isDisable = !ModInstaller.hasBranch()
    }

    @FXML
    private fun connectToRepository() {
        MainController.executeTask(
            "Loading repository..."
        ) { taskUpdateStage: Stage ->
            object : Task<Unit>() {
                override fun call() {
                    var repoURL: String
                    val errorHandler = TextHandler {
                        Platform.runLater {
                            val msg = ResizableAlert(AlertType.ERROR, it)
                            msg.show()
                        }
                    }

                    if (ModInstaller.connectToRepository(
                            urlField!!.text,
                            { updateMessage(it) },
                            errorHandler
                        ).also { repoURL = it!! } == null
                    ) return

                    updateMessage("Fetching branch list...")
                    if (!getBranches(errorHandler)) return

                    updateMessage("Getting mod description...")
                    if (!getDescription(errorHandler)) return

                    Platform.runLater {
                        setUrlText("Connected to $repoURL")
                        try {
                            Recommendations.instance.addItem(urlField!!.text)
                            recommendationList!!.items.setAll(Recommendations.instance.recommendationList)
                        } catch (e: IOException) {
                            val msg = Alert(AlertType.WARNING, "Could not save history.json! Cause: $e")
                            e.printStackTrace()
                            msg.showAndWait()
                        }
                    }
                }

                override fun succeeded() {
                    taskUpdateStage.close()
                }
            }
        }
    }

    @FXML
    private fun installMod() {
        MainController.executeTask(
            "Preparing to install mod..."
        ) { taskInstallStage: Stage ->
            object : Task<Unit>() {
                override fun call() {
                    updateMessage("Preparing to install mod...")
                    ModInstaller.installMod(
                        { warning ->
                            Platform.runLater {
                                val msg: Alert = ResizableAlert(AlertType.WARNING, warning)
                                msg.show()
                            }
                        },
                        { message -> updateMessage(message) },
                        { info ->
                            Platform.runLater {
                                val msg: Alert = ResizableAlert(AlertType.INFORMATION, info)
                                msg.show()
                            }
                        },
                        { error ->
                            Platform.runLater {
                                val msg: Alert = ResizableAlert(AlertType.ERROR, error)
                                msg.show()
                            }
                        }
                    )
                    return
                }

                override fun succeeded() {
                    taskInstallStage.close()
                }
            }
        }
    }

    fun getBranches(errorHandler: TextHandler?): Boolean {
        val tagNames = ModInstaller.getBranches(errorHandler) ?: return false
        Platform.runLater {
            branchList!!.items.clear()
            branchList!!.items.addAll(tagNames)
        }
        return true
    }

    fun getDescription(errorHandler: TextHandler?): Boolean {
        val description = ModInstaller.getDescription(errorHandler) ?: return false
        Platform.runLater {
            modInfo!!.text = description
            branchInfo!!.text = "No branches selected"
        }
        return true
    }

    fun setUrlText(url: String?) {
        urlLabel!!.text = url
    }

    fun deleteRepository() {
        val dialog = ResizableAlert(
            AlertType.CONFIRMATION,
            "Are you sure you want to delete this repository from your computer? You will not be able to access it again until you redownload it."
        )
        val dlgButtons = dialog.dialogPane.buttonTypes
        dlgButtons.clear()
        dlgButtons.add(ButtonType.YES)
        dlgButtons.add(ButtonType.NO)
        dialog.headerText = "Delete Repository?"
        dialog.showAndWait()
            .filter { response: ButtonType -> response == ButtonType.YES }
            .ifPresent {
                if (ModInstaller.deleteRepository { error: String? ->
                        val msg = ResizableAlert(AlertType.ERROR, error)
                        msg.show()
                    }) {
                    branchList!!.items.clear()
                    modInfo!!.text = ""
                    branchInfo!!.text = ""
                }
            }
    }

    fun uninstallMod() {
        val dialog = ResizableAlert(
            AlertType.CONFIRMATION,
            "Are you sure you want to uninstall this mod? The associated repository will not be removed from your computer."
        )
        val dlgButtons = dialog.dialogPane.buttonTypes
        dlgButtons.clear()
        dlgButtons.add(ButtonType.YES)
        dlgButtons.add(ButtonType.NO)
        dialog.headerText = "Uninstall Mod?"
        dialog.showAndWait()
            .filter { response: ButtonType -> response == ButtonType.YES }
            .ifPresent {
                ModInstaller.uninstallMod(
                    { error: String? ->
                        val msg = ResizableAlert(AlertType.ERROR, error)
                        msg.show()
                    },
                    modList!!.selectionModel.selectedItem
                )
            }
    }
}